### PR TITLE
RISC-V

### PR DESCRIPTION
**Pull request by nastya-nizharadze**
add interrupt handler, exception handler, drivers for interrupt and timer, and template for sifive-u

### CODE REVIEW

File: src/arch/riscv/include/asm/entry.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+		lw      t6, PT_MSTATUS(sp)
+		csrw    mstatus, t6
+
+		lw      t6, PT_PC(sp)
+		csrw    mepc, t6
+
+		lw      t6, PT_R31(sp)
+
+		//lw      sp, PT_SP(sp)
```
Comment (anton-bondarev) [inline src/arch/riscv/include/asm/entry.h]: rm this please

File: src/arch/riscv/kernel/Mybuild [hunk 2]
```diff
@@ -2,17 +2,23 @@ package embox.arch.riscv.kernel
...
 
 module interrupt extends embox.arch.interrupt {
-	source "ipl_impl.c", "ipl_impl.h"
+	source "interrupt.c", "ipl_impl.c", "ipl_impl.h"
+	depends embox.kernel.irq
 }
+
+module entry extends embox.arch.entry {
+	source "entry.S"
+}
```
Comment (anton-bondarev) [inline src/arch/riscv/kernel/Mybuild]: \n please

File: src/arch/riscv/kernel/entry.S [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+.align 4
+.global trap_handler
+trap_handler:
+
+	SAVE_ALL
+	csrr    t6, mcause
+	srli    t6, t6, 31
+	beqz    t6, riscv_exception_handler
+	mv      a0, sp
```
Comment (anton-bondarev) [inline src/arch/riscv/kernel/entry.S]: sp + PT_SIZE?

File: src/arch/riscv/include/asm/entry.h [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+#ifndef RISCV_ENTRY_H_
+#define RISCV_ENTRY_H_
+
+#include <asm/ptrace.h>
+
+#ifndef __ASSEMBLER__
+
+#else /* __ASSEMBLER__ */
+
+	.macro SAVE_ALL
```
Comment (alexkalmuk) [inline src/arch/riscv/include/asm/entry.h]: Can you please provide a link in the comment which described this "mixed" registers order?

Comment (nastya-nizharadze) [reply src/arch/riscv/include/asm/entry.h]: https://riscv.org/specifications/ page 137

Comment (nastya-nizharadze) [reply src/arch/riscv/include/asm/entry.h]: x0 | Zero 
x1 | ra 
x2 | sp
x3 | gp 
x4 | tp 
x5 | t0 
x6–7 | t1–2 
x8 | s0/fp 
x9 | s1 
x10–11 | a0–1 
x12–17 | a2–7
x18–27 | s2–11
x28–31 | t3–6

File: src/arch/riscv/include/asm/entry.h (hunk 1 above)
Comment (nastya-nizharadze) [reply src/arch/riscv/include/asm/entry.h]: Done

File: src/arch/riscv/kernel/Mybuild (hunk 2 above)
Comment (nastya-nizharadze) [reply src/arch/riscv/kernel/Mybuild]: Done

File: src/drivers/clock/Mybuild [hunk 5]
```diff
@@ -69,6 +69,18 @@ module mips_clk extends embox.arch.clock {
...
+module riscv_clk extends embox.arch.clock {
+	source "riscv_clk.c"
+
+	depends embox.kernel.irq
+	depends embox.kernel.time.clock_source
+	depends embox.arch.interrupt
+	@NoRuntime depends embox.kernel.time.timer_handler
+
+	option number base_mtime=0x200bff8
+	option number base_mtimecmp=0x2004000
```
Comment (alexkalmuk) [inline src/drivers/clock/Mybuild]: Can you please provide document for this?

Comment (nastya-nizharadze) [reply src/drivers/clock/Mybuild]: https://sifive.cdn.prismic.io/sifive%2Fc89f6e5a-cf9e-44c3-a3db-04420702dcc1_sifive+e31+manual+v19.08.pdf
page 24
6.1 CLINT Memory Map

Comment (alexkalmuk) [reply src/drivers/clock/Mybuild]: Thank you.

File: src/arch/riscv/include/asm/ptrace.h [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	unsigned int pc;
+} pt_regs_t;
+
+static inline void ptregs_retcode(struct pt_regs *ptregs, int retcode) {
+	ptregs->reg[5] = retcode; /* $a0 ($10) */
+}
+
+#else /* __ASSEMBLER__ */
+
+
```
Comment (Deryugin) [inline src/arch/riscv/include/asm/ptrace.h]: Please, remove double blank line

File: src/arch/riscv/include/asm/ptrace.h [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+#define PT_MSTATUS 0x7C
+#define PT_PC      0x80
+
+/* sizeof(pt_regs) */
+#define PT_SIZE    0x84
+
+#endif /* __ASSEMBLER__ */
+
+
```
Comment (Deryugin) [inline src/arch/riscv/include/asm/ptrace.h]: Please, remove double blank line

**Comment by anton-bondarev** [comment]
work on #1047

**Pull request merged**